hnn_core/feed.py

```diff
@@ -35,252 +74,238 @@ class ExtFeed(object):
         Parameters of the external input feed, arranged into a dictionary.
     gid : int
         The cell ID.
+    """
+    prng, prng2 = _get_prng(
+        seed=params['prng_seedcore'],
+        gid=gid,
+        sync_evinput=params.get('sync_evinput', False))
 
-    Attributes
+    # check feed name validity, allowing substring matches ('evprox1' etc)
+    valid_feeds = ['extpois', 'extgauss', 'common', 'evprox', 'evdist']
+    # NB check if feed_type has a valid substring, not vice versa
+    matches = [f for f in valid_feeds if f in feed_type]
+    if len(matches) == 0:
+        raise ValueError('Invalid external feed: %s' % feed_type)
+    elif len(matches) > 1:
+        raise ValueError('Ambiguous external feed: %s' % feed_type)
+
+    # Return values not checked: False if all weights for given feed type
+    # are zero. Designed to be silent so that zeroing input weights
+    # effectively disables each.
+    zero_ampa_nmda = False
+    if target_cell_type in params:
+        zero_ampa_nmda = (params[target_cell_type][0] <= 0.0 and
+                          params[target_cell_type][1] <= 0.0)
+
+    all_syn_weights_zero = True
+    for key in params.keys():
+        if key.startswith(('L2Pyr', 'L5Pyr', 'L2Bask', 'L5Bask')):
+            if params[key][0] > 0.0:
+                all_syn_weights_zero = False
+
+    event_times = list()
+    if feed_type == 'extpois' and not zero_ampa_nmda:
+        event_times = _create_extpois(
+            t0=params['t_interval'][0],
+            T=params['t_interval'][1],
+            # ind 3 is frequency (lamtha))
+            lamtha=params[target_cell_type][3],
+            prng=prng)
+    elif feed_type.startswith(('evprox', 'evdist')) and \
+            target_cell_type in params:
+        event_times = _create_evoked(
+            mu=params['t0'],
+            # ind 3 is sigma_t (stdev))
+            sigma=params[target_cell_type][3],
+            numspikes=int(params['numspikes']),
+            prng=prng)
+    elif feed_type == 'extgauss' and not zero_ampa_nmda:
+        event_times = _create_extgauss(
+            mu=params[target_cell_type][3],
+            sigma=params[target_cell_type][4],
+            prng=prng)
+    elif feed_type == 'common' and not all_syn_weights_zero:
+        event_times = _create_common_input(
+            distribution=params['distribution'],
+            t0=params['t0'],
+            t0_stdev=params['t0_stdev'],
+            tstop=params['tstop'],
+            f_input=params['f_input'],
+            stdev=params['stdev'],
+            repeats=params['repeats'],
+            events_per_cycle=params['events_per_cycle'],
+            prng=prng,
+            prng2=prng2)
+
+    # brute force remove non-zero times. Might result in fewer vals
+    # than desired
+    # values MUST be sorted for VecStim()!
+    if len(event_times) > 0:
+        event_times = event_times[event_times > 0]
+        event_times.sort()
+        event_times.tolist()
```

Comment:
Why is this necessary? Shouldn't this be enforced when `event_times` is created?

Reply:
The numpy functions return arrays, so you need to convert them to a list

Reply:
My question was more about why `event_times > 0` isn't enforced from the beginning, but I think I get it now.

Reply:
Actually it's a very good point now that you mention it. Only the poisson inputs are guaranteed to have `event_times > 0`. For the rest, it depends. For example, if it's Gaussian and the mean is too close to 0., then some events in the beginning will be truncated. Honestly, I don't get the logic of that. If you remove some events, then what you get is actually a [truncated Gaussian distribution](https://en.wikipedia.org/wiki/Truncated_normal_distribution) not a Gaussian distribution.